addons/account/models/account_move.py

```diff
@@ -2415,29 +2415,31 @@ def _apply_cash_rounding(self, diff_balance, diff_amount_currency, cash_rounding
 
     @contextmanager
     def _sync_unbalanced_lines(self, container):
+        def has_tax(move):
+            return bool(move.line_ids.tax_ids)
+
+        move_had_tax = {move: has_tax(move) for move in container['records']}
         yield
+        move_has_tax = {move: has_tax(move) for move in container['records']}
         # Skip posted moves.
-        for invoice in (x for x in container['records'] if x.state != 'posted'):
-
-            # Unlink tax lines if all taxes have been removed.
-            if not invoice.line_ids.tax_ids:
-                # if there isn't any tax but there remains a tax_line_id, it means we are currently in the process of
-                # removing the taxes from the entry. Thus, we want the automatic balancing to happen in order  to have
-                # a smooth process for tax deletion
-                if not invoice.line_ids.filtered('tax_line_id'):
-                    continue
-                invoice.line_ids.filtered('tax_line_id').unlink()
+        for move in (x for x in container['records'] if x.state != 'posted'):
```

Comment:
leaving the loop variable as `invoice` instead of `move` would eliminate some unnecessary diff

Reply:
It is very confusing because it will never be an invoice.